feat(config): Allow Scoop to ignore running processes during reset/uninstall/update

CrendKing opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->
<!-- To help with semantic versioning the PR title should start with one of the conventional commit types. -->
<!-- The conventional commit types for Semantic PR are: feat, fix, docs, style, refactor, perf, test, build, ci, chore, revert -->

<!--
  By opening this PR you confirm that you have searched for similar issues/PRs here already.
  Failing to do so will most likely result in closing of this PR without any explanation.
  It is also mandatory to open a relevant issue for discussion with the maintainers,
  before creating any new PR.
  Read the contributing guide first to save both your and our time.
-->

#### Description
<!-- Describe your changes in detail -->
When any process of an application is still running, currently during uninstallation (`uninstall` or `update`), it unconditionally prints error message "Application is still running. Close all instances and try again" and exits. While this approach minimizes risk of data loss, manifest author has no control over the procedure.

Some apps consist of constantly running services and drivers, which has no GUI. In such case, user may have no idea how to "close all instances", and get stuck. Manifest author might be able to `Stop-Service` or call certain external command to stop those processes, but the current logic prevents this.

One simple way to fix this is adding a prompt during uninstallation, as this PR demonstrates. If Scoop detects process running, it calls `Read-Host` to ask user permission to proceed the "uninstaller" script, assuming which will take care of the process termination. In case user answers "N", there will be no change.

Currently there is no flag to guard this change, so all users will be immediately affected by this change. I have no previous experience on how Scoop usually handle these kind of changes. We could create a new property in manifest structure to activate this.

Furthermore, if this is approved, we could even always attempt to call `Stop-Process -Name $bin` after prompt, where "$bin" is the string provided in the manifest's "bin" property, if exist. This way we could streamline the app update even more.

#### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
Closes #4710

#### How Has This Been Tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->
1. Install an app.
2. Launch the app without terminating it.
3. `scoop uninstall <app_name>`.
4. Observe a prompt shows during uninstallation.
4.1. Press any key other than "y" results in the "Application is still running. Close all instances and try again" message.
4.2. Press "y", the uninstallation continues.
5. If <app> manifest has uninstaller script that `Stop-Process -Name <app_name>`, app is closed and unininstalled successfully.
5.1. If not, uninstallation fails with error "Couldn't remove <app_dir>; it may be in use."

#### Checklist:
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have read the [Contributing Guide](https://github.com/ScoopInstaller/.github/blob/main/.github/CONTRIBUTING.md).
- [x] I have updated the documentation accordingly.
- [x] I have updated the tests accordingly.

tech189 commented on libexec/scoop-reset.ps1:
```diff
@@ -69,7 +69,8 @@ $apps | ForEach-Object {
     $processdir = $dir | Select-Object -ExpandProperty Path
-    if (Get-Process | Where-Object { $_.Path -like "$processdir\*" }) {
+    if ((Get-Process | Where-Object { $_.Path -like "$processdir\*" }) -and
+        (Read-Host "Application '$app' is still running. Do you want to proceed uninstallation? All unsaved data will be lost [y/N]") -ne 'y') {
```
```suggestion
        (Read-Host "Application '$app' is still running. Do you want to proceed with the reset? All unsaved data will be lost [y/N]") -ne 'y') {
```
Not sure if this prompt is necessary, will any data actually be lost - doesn't reset just recreate shims and shortcuts?

CrendKing replied:
The "data loss" I meant is in-memory progress, like if you are editing a file without saving to disk. Nothing already on disk is not lost.

What's your suggestion?

tech189 replied:
I don't think scoop stops the process when resetting. Have a look at lines 78-90 of `scoop-reset.ps1`, it's just recreating shims, start menu shortcuts, re-adding environment variables and any directories to the PATH variable, plus recreating the junctions for the `current` folder and the persisted folders. You can try for yourself, I just reset audacity while it was running and it wasn't stopped. It still remembers my preferences so persisted data was not lost either.

So if that's true my suggestion is remove this prompt for reset but keep it for update/uninstallation. (see my other suggestions below for better wording in update and uninstallation)

CrendKing replied:
Oh, you are talking specifically for the "reset" case. Sorry I misunderstood.

So the reason why you don't see the "stop process" message and the new prompt, I think, is because the current `scoop-reset.ps1` has a problem. Instead of getting the `$processdir = appdir $app $global | Resolve-Path | Select-Object -ExpandProperty Path` like `scoop-uninstall.ps1`, it's doing `$processdir = $dir | Select-Object -ExpandProperty Path`. This means the variable points to the versioned directory instead of the `current` junction. Since when you launch an app the current directory is most likely the `current` junction, that process running check in `scoop-reset.ps1` never worked. I just noticed this bug.

Once you change line 71 to the same as `scoop-uninstall.ps1`, you will see the prompt. Can you confirm?

tech189 replied:
I had a look at #4310, yes it looks like this prompt is necessary. 

>Once you change line 71 to the same as scoop-uninstall.ps1, you will see the prompt. Can you confirm?

Yes I see the prompt.

In that case, my only suggestion for this file is change the wording from 'proceed uninstallation?' to 'proceed with the reset?'

CrendKing replied:
Yes. I'll change the wording if this approach is approved.

tech189 commented on libexec/scoop-uninstall.ps1:
```diff
@@ -60,7 +60,8 @@ if (!$apps) { exit 0 }
         $processdir = $appDir | Resolve-Path | Select-Object -ExpandProperty Path
-        if (Get-Process | Where-Object { $_.Path -like "$processdir\*" }) {
+        if ((Get-Process | Where-Object { $_.Path -like "$processdir\*" }) -and
+            (Read-Host "Application '$app' is still running. Do you want to proceed uninstallation? All unsaved data will be lost [y/N]") -ne 'y') {
```
```suggestion
            (Read-Host "Application '$app' is still running. Do you want to proceed with uninstallation? All unsaved data will be lost [y/N]") -ne 'y') {
```

tech189 commented on libexec/scoop-update.ps1:
```diff
@@ -232,7 +232,8 @@ function update($app, $global, $quiet = $false, $independent, $suggested, $use_c
     $processdir = appdir $app $global | Resolve-Path | Select-Object -ExpandProperty Path
-    if (Get-Process | Where-Object { $_.Path -like "$processdir\*" }) {
+    if ((Get-Process | Where-Object { $_.Path -like "$processdir\*" }) -and
+        (Read-Host "Application '$app' is still running. Do you want to proceed uninstallation? All unsaved data will be lost [y/N]") -ne 'y') {
```
```suggestion
        (Read-Host "Application '$app' is still running. Do you want to proceed with the update? All unsaved data will be lost [y/N]") -ne 'y') {
```

tech189 reviewed (commented):
Just some suggestions for improving the wording, I haven't actually tested the functionality - I'll leave that to the maintainers.

niheaven commented:
I don't like this idea, a conclusion notes after all uninstallation to remind user that some apps are still running and not uninstalled is more better, since this one will interrupt the unattended uninstalling process.

Or add a new config option to let scoop force exit all running processes and uninstall/update the apps.

CrendKing commented:
I agree that this change is backward incompatible and breaks unattended process. As I mentioned in the opening, if the community would come up with a "force exit running process" option, I'm OK with that too. Maybe provide that as a `scoop config` option (i.e. when true, always force exit instead of always error and fail uninstall/reset).

@rashil2000, what do you think?

rashil2000 commented:
Hmm, the config option is a much better UX. It does not break any previous user workflows, and offers customizability without needing to ask the user for a prompt each time. 

For implementation of the config option, you can refer to this PR - #4155 

You'll also need to update the libexec/scoop-config.ps1 file, for documentation.

CrendKing commented:
OK. Just to recap the new requirements:

1. Add a new boolean config to the `scoop config` command, which enables the "force kill processes" function. This defaults to false for backward compatibility.
2. For uninstall, update and reset, if Scoop detects any process running from the app directory, AND the config above is enabled by user, instead of print error message and exit, forcefully kill all running process by calling `Stop-Process`. Print relevant message for this action.
3. Afterwards, proceed the uninstall/update/reset procedure as usual.

Is this good?

CrendKing commented:
Updated the PR. Moved the handling logic to core.ps1 for reuse between all the 3 use cases. If this function is better placed in a different file, please advise.

rashil2000 commented:
Do you think Scoop should stop the process if enabled? It will most likely fail (and print a wall of errors) - unless they are stopped in the proper order and with the proper flags (like `-Force` if necessary). I think we should just let the user stop the processes themselves. 

The config option should be boolean - false (by default) to retain current behaviour, and true to just print the warning and proceed with uninstaller script.

niheaven commented:
For the code, put the function under `install.ps1` IMO.

CrendKing commented:
> It will most likely fail (and print a wall of errors) - unless they are stopped in the proper order and with the proper flags (like -Force if necessary)

According to the [PowerShell doc](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.management/stop-process?view=powershell-7.2) without `-Force` the cmdlet doesn't stop process spawned from different user account. So I suppose since the function could be thrown with a global app, which could be started by other user, we should use `-Force` here. Gotcha.

I tried several apps locally, and I don't get any problem. If you have see errors, could you provide the repro steps?

> true to just print the warning and proceed with uninstaller script.

To clarify, I think we are discussing about two use cases:

1. Normal everyday app is running when doing update/uninstall, where a generic `Stop-Process` is sufficient to handle.
2. Special app (such as Everything driver/service from https://github.com/ScoopInstaller/Scoop/issues/4710), which no GUI app is running. Stopping those processes require special tools/steps.

For 2, I completely agree we should leave to the uninstaller script.
For 1, if we just display a warning, since most manifest today does not have `Stop-Process`, it's not gonna change anything. Besides, what other special ways can the manifest authors do better than a generic `Stop-Process`?

> For the code, put the function under install.ps1 IMO.

Gotcha.

rashil2000 commented:
> If you have see errors with this impl, could you provide the repro steps?

I've encountered problems outside of Scoop, where a generic Stop-Process isn't enough.

> For 1, if we just display a warning, since most manifest today does not have Stop-Process, it's not gonna change anything.

That's preferable, to be honest. Scoop is a pretty non-invasive manager, and doing something outside of its scope will deviate from that. As I said, if any apps need special-casing, manifest authors will include a Stop-Process in their uninstaller scripts, and I think that's sufficient.

CrendKing commented:
Understood. I'll remove the `Stop-Process` and update.

CrendKing commented:
Manual test steps I did:

1. Install an app A, which has GUI. Config JSON is empty.
2. Launch A. Leave it running. Do `scoop uninstall A`.
3. Observe error message `ERROR Application "A" is still running. Close all instances and try again.` and Scoop terminates there.
4. Add config `"ignoreRunning": false,` and repeat steps 2-3. Observe exactly the same result.
5. Change `ignoreRunning` to `true`. Observe warn message `WARN  Application "A" is still running. Scoop is configured to ignore this condition.` instead and other messages such as "Removing shim". Scoop stops with error `ERROR Couldn't remove 'D:\Scoop\apps\A\0.0.1'; it may be in use.` This is because A manifest does not handle running process in uninstaller.
6. Reinstall A. Add the following script to A\0.0.1\manifest.json:
```
    "uninstaller": {
        "script": "Stop-Process -Name A"
    },
```
7. Do uninstallation. Observe the warning message, then `Running uninstaller script...` , A's process gets killed, and eventually `'A' was uninstalled.`

=====

Is there any test file need update?

rashil2000 commented:
Looks great. I'll do a proper review later today.

Hmm, I don't think there are tests for this. @niheaven can you confirm?

rashil2000 commented:
Vague idea: instead of a config option, how about we detect the presence of `uninstaller` field in the manifest, to automate this process? If `uninstaller` exists, Scoop will warn the user and proceed. If it doesn't, it pops up an error and stops.

niheaven commented:
Nope, this process related to specific app installation/update cannot be tested. Just run `.\bin\test.ps1` to see if all passes.

BTW, you should check config before running the new function to avoid `Get-Process` for every uninstallation/updating.

niheaven commented:
Aha sorry, I misunderstood the logic of the function 🤣 Please ignore it.

rashil2000 reviewed (changes requested):
We should call it `ignore_running_processes` to be more clear.

rashil2000 commented:
Please add an entry in the changelog

niheaven commented:
Call the function `ignore_running_process` and flip the output (`$true` for ignore/`$false` for break) to make this more readable.

CrendKing commented:
Oh, you mean I should make the function "ignore_running_process", not "handle_running_processes"? What about the config? Any suggestion?

Sorry for misunderstanding.

rashil2000 commented:
This should be `feat` instead of `fix`

niheaven commented:
It should be `stop_running_process` 🤣 

> Or you can change the PR yourself, feel free to just modify it.

Okay, I'll change it a little, feel free to reverse my commit if it got something wrong

niheaven commented:
![image](https://user-images.githubusercontent.com/5832170/153535238-c2032024-ca1c-4e8b-8418-499e498cfbfa.png)

niheaven commented:
@CrendKing If this is fine, I'll merge it.

CrendKing commented:
Of course. Go ahead.

rashil2000 commented:
Thanks for the contribution @CrendKing!

CrendKing commented:
@niheaven I think https://github.com/ScoopInstaller/Scoop/pull/4713/commits/e55639d2d7913ef5e5d8082ef976825e08aea094 introduced a bug. Previously either the error or warn message would print only if there is some running process. After that commit, the warn message would print as long as that config is set.

Should I make a PR to correct it or you can do it?

niheaven commented:
It's been fixed in #4731, you could try it 🤣

CrendKing commented:
That's exactly what I'm looking for. Sorry didn't notice that commit. Thank you!
